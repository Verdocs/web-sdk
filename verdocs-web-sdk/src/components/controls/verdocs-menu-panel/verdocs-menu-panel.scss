@import '../../../theme.scss';

#verdocs-menu-panel-overlay {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 10000;
  position: fixed;
  background-color: #0000007f;
}

#verdocs-menu-panel {
  opacity: 0;
  z-index: 10001;
  position: fixed;
  overflow-y: scroll;
  background-color: #ffffff;
  font-family: $verdocs-primary-font;

  -webkit-animation: showVerdocsMenuPanel 0.35s;
  -webkit-animation-fill-mode: forwards;
}

// TODO: Parameterize the width
@-webkit-keyframes showVerdocsMenuPanel {
  0% {
    opacity: 0;
    right: -300px;
  }

  100% {
    right: 0;
    opacity: 1;
  }
}
