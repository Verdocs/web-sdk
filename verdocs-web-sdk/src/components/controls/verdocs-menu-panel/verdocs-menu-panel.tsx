import {Component, EventEmitter, h, Host, Listen, Prop, Event} from '@stencil/core';

/**
 * Display a menu panel in a left or right sidebar. The panel will animate (slide)
 * as it appears, and an background will be shown over the rest of the page. If
 * the background overlay is present, it can be clicked to dismiss the panel.
 *
 * ```ts
 * <verdocs-menu-panel>
 *   <div style="padding: 20px;">Menu Panel</div>
 * </verdocs-menu-panel>
 * ```
 */
@Component({
  tag: 'verdocs-menu-panel',
  styleUrl: 'verdocs-menu-panel.scss',
  shadow: false,
})
export class VerdocsMenuPanel {
  private portal: HTMLElement;
  private element: HTMLElement;
  private moved: boolean = false;
  private overlayId = `verdocs-menu-panel-overlay`;
  private panelId = `verdocs-menu-panel`;

  /**
   * Which side of the screen to place the panel.
   */
  @Prop() side: 'left' | 'right' = 'right';

  /**
   * Whether to show an overlay over the rest of the page.
   */
  @Prop() overlay = true;

  /**
   * The width of the panel.
   */
  @Prop() width = 300;

  @Event() close: EventEmitter<void>;

  @Listen('click', {target: 'document'})
  handleClick(e: any) {
    if (!this.element.contains(e.target)) {
      this.hide();
      this.close?.emit();
    }
  }

  componentWillLoad() {
    this.hide();

    if (this.overlay) {
      this.portal = document.createElement('div');
      this.portal.setAttribute('id', this.overlayId);
      document.body.append(this.portal);
    }

    this.portal = document.createElement('div');
    this.portal.setAttribute('id', this.panelId);
    this.portal.style.top = `0px`;
    this.portal.style.bottom = `0px`;
    this.portal.style.width = `${this.width}px`;

    if (this.side === 'left') {
      this.portal.style.left = `0px`;
    } else {
      this.portal.style.right = `0px`;
    }

    document.body.append(this.portal);
  }

  componentDidLoad() {
    this.portal.appendChild(this.element);
  }

  disconnectedCallback() {
    this.moved ? this.portal.remove() : (this.moved = true);
  }

  hide() {
    document.getElementById(this.overlayId)?.remove();
    document.getElementById(this.panelId)?.remove();
  }

  render() {
    return (
      <Host ref={el => (this.element = el)}>
        <slot />
      </Host>
    );
  }
}
