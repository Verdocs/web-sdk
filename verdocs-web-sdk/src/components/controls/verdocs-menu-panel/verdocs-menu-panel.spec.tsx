import { newSpecPage } from '@stencil/core/testing';
import { VerdocsMenuPanel } from './verdocs-menu-panel';

describe('verdocs-menu-panel', () => {
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsMenuPanel],
      html: `<verdocs-menu-panel></verdocs-menu-panel>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with side left and custom width', async () => {
    const page = await newSpecPage({
      components: [VerdocsMenuPanel],
      html: `<verdocs-menu-panel side="left" width="400"></verdocs-menu-panel>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.side).toBe('left');
    expect(page.rootInstance.width).toBe(400);
  });

  it('renders with overlay disabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsMenuPanel],
      html: `<verdocs-menu-panel overlay="false"></verdocs-menu-panel>`,
    });
    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.overlay).toBe(false);
  });

  // Note: Simulating document click to test close event may not work in JSDOM, but we can check event emission logic.
  it('emits close event', async () => {
    const page = await newSpecPage({
      components: [VerdocsMenuPanel],
      html: `<verdocs-menu-panel></verdocs-menu-panel>`,
    });
    const spy = jest.fn();
    page.root.addEventListener('close', spy);
    // Directly call the handler to simulate
    (page.rootInstance as any).handleClick({ target: document.createElement('div'), stopPropagation: () => {} });
    expect(spy).toHaveBeenCalled();
  });
});
