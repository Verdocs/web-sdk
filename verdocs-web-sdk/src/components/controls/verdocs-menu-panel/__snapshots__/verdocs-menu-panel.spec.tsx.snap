// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-menu-panel renders with default props 1`] = `
<verdocs-menu-panel>
  <!---->
</verdocs-menu-panel>
`;

exports[`verdocs-menu-panel renders with overlay disabled 1`] = `
<verdocs-menu-panel overlay="false">
  <!---->
</verdocs-menu-panel>
`;

exports[`verdocs-menu-panel renders with side left and custom width 1`] = `
<verdocs-menu-panel side="left" width="400">
  <!---->
</verdocs-menu-panel>
`;
